### DISCUSSION TITLE
Best practice to troubleshoot Heap usage

### DISCUSSION THREAD

**Discussion by lorezyra** [hexojs/hexo#5157] [Question / Answer] [answered]
Other than using the VScode extension [Flame Chart Visualizer for JavaScript Profiles](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-js-profile-flame), how should one go about reviewing the amount of memory Hexo consumes?

I have over [1400+ post](https://Blog.RichieBartlett.com) (over 9400 files) that take about 70sec to generate but consumes over 14GB(!) of memory. When I use the default theme, it's able to process everything under 4GB. Total folder size of all assets is less than 4GB. I'd like to figure out why it takes another 10GB of memory under the new theme I'm building.

Not just troubleshooting Hexo, but how to troubleshoot EJS files? The debugger doesn't exactly pause anywhere within the EJS files (as it does for the node_modules JS files...).

## Reference
- https://blog.richiebartlett.com/Work_%E4%BB%95%E4%BA%8B/Programming_%E9%96%8B%E7%99%BA/Debugging-HEXO-in-VScode/
- https://blog.richiebartlett.com/Work_%E4%BB%95%E4%BA%8B/Setup-VScode-for-Hexo/

## Related
- https://github.com/hexojs/hexo/issues/3298
- https://github.com/hexojs/hexo-server/issues/275
- https://github.com/hexojs/hexo/blob/master/.github/CONTRIBUTING.md

**Comment by SukkaW** [comment]
It sounds like your theme is leaking memory.

You can use this tool to debug any performance issue (including memory leak): https://clinicjs.org/

**Reply by lorezyra** [reply]
I'll check it out. Although, it would be nice to do everything within VScode. I'm wondering how the HEXO team does their performance testing.

**Reply by SukkaW** [reply]
> I'm wondering how the HEXO team does their performance testing.

https://github.com/hexojs/hexo/blob/master/test/benchmark.js

**Reply by lorezyra** [reply]
> It sounds like your theme is leaking memory.

Indeed. Seems my `donate.ejs` _partial is taking considerable memory on my post pages. That file has my SVG images embedded for crypto payments. Now that I've a lot of posts, it makes more sense to remove the embedded SVG images and have them as separate files. I still need to dig deeper and make my theme more performant while still retaining all its features...

**Comment by lorezyra** [comment, marked as answer]
I figured out the issues I had with VScode not attaching to the nodeJS process running HEXO. Now when I run the (vscode) debugger with the FlameChart extension, I can see the active heap usage along with CPU consumption.

Here's the VScode `launch.json` config:

```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "type": "node",
            "request": "launch",
            "name": "HEXO server",
            "program": "${workspaceFolder}/node_modules/hexo-cli/bin/hexo",
            "args": [
                "server",
                "-p 4000"
            ],
            "restart": true,
            "runtimeExecutable": "node",
            "runtimeArgs": [
                "--inspect"
            ],
            "stopOnEntry": true,
            "sourceMaps": true,
            "cwd": "${workspaceFolder}"
        },
        {
            "type": "node",
            "request": "launch",
            "name": "Hexo generate",
            "program": "${workspaceFolder}/node_modules/hexo-cli/bin/hexo",
            "args": [
                "generate"
            ],
            "sourceMaps": true,
            "cwd": "${workspaceFolder}"
        }
    ]
}
```

📝 I'm also updating my blog on how to debug Hexo themes and plugins...
